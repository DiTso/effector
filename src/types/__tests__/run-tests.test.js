// @flow

import execa from 'execa'

jest.setTimeout(40000)

test('TypeScript', async() => {
  try {
    const data = await execa('npx', ['tsc', '-p', 'src/types'])
    expect(data).toMatchSnapshot('resolved')
  } catch (err) {
    expect(err.message).toMatchSnapshot('rejected')
  }
})

test('Flow', async() => {
  try {
    const data = await execa('npx', [
      'flow',
      'focus-check',
      '--strip-root',
      '--json',
      'src/types/types.test.js',
    ])
    expect(JSON.parse(data.stdout)).toMatchSnapshot('resolved')
  } catch (err) {
    const data = JSON.parse(
      err.message.substring(err.message.indexOf('\n') + 1).trim(),
    )
    expect(data).toMatchSnapshot('json rejected')
    expect(
      data.errors.map(error => error.message.map(p => p.descr)),
    ).toMatchSnapshot('json messages')
  }
})
