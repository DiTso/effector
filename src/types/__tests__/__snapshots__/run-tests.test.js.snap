// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow: json messages 1`] = `
Array [
  Array [
    "Cannot assign \`c\` to \`check2\` because number [1] is incompatible with string [2] in type argument \`Payload\` [3].",
  ],
  Array [
    "Cannot assign \`e\` to \`check4\` because number [1] is incompatible with string [2] in type argument \`State\` [3].",
  ],
  Array [
    "Cannot assign \`g\` to \`check6\` because string [1] is incompatible with number [2] in type argument \`Payload\` [3].",
  ],
  Array [
    "Cannot assign \`computed\` to \`check2\` because string [1] is incompatible with number [2] in type argument \`Payload\` [3].",
  ],
  Array [
    "Cannot assign \`createStore(...)\` to \`store2\` because number [1] is incompatible with string [2] in type argument \`State\` [3].",
  ],
  Array [
    "Cannot assign \`computed\` to \`check2\` because string [1] is incompatible with number [2] in type argument \`State\` [3].",
  ],
  Array [
    "Cannot call \`createDomain\` with \`234\` bound to \`domainName\` because number [1] is incompatible with string [2].",
  ],
  Array [
    "Cannot call \`createDomain\` with object literal bound to \`domainName\` because object literal [1] is incompatible with string [2].",
  ],
]
`;

exports[`Flow: json rejected 1`] = `
Object {
  "errors": Array [
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const a = createEvent<number>()",
              "descr": "[1]",
              "end": 32,
              "endline": 23,
              "line": 23,
              "loc": Object {
                "end": Object {
                  "column": 32,
                  "line": 23,
                  "offset": 366,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 27,
                  "line": 23,
                  "offset": 360,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 27,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check2: Event<string> = c",
              "descr": "[2]",
              "end": 30,
              "endline": 28,
              "line": 28,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 28,
                  "offset": 501,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 28,
                  "offset": 495,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Event<Payload> implements Unit<Payload> {",
              "descr": "[3]",
              "end": 34,
              "endline": 29,
              "line": 29,
              "loc": Object {
                "end": Object {
                  "column": 34,
                  "line": 29,
                  "offset": 486,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 29,
                  "offset": 479,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check2: Event<string> = c",
          "descr": "Cannot assign \`c\` to \`check2\` because number [1] is incompatible with string [2] in type argument \`Payload\` [3].",
          "end": 35,
          "endline": 28,
          "line": 28,
          "loc": Object {
            "end": Object {
              "column": 35,
              "line": 28,
              "offset": 506,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 28,
              "offset": 505,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check3: Store<number> = e",
              "descr": "[1]",
              "end": 30,
              "endline": 33,
              "line": 33,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 33,
                  "offset": 595,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 33,
                  "offset": 589,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check4: Store<string> = e",
              "descr": "[2]",
              "end": 30,
              "endline": 34,
              "line": 34,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 34,
                  "offset": 631,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 34,
                  "offset": 625,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Store<State> implements Unit<State> {",
              "descr": "[3]",
              "end": 32,
              "endline": 69,
              "line": 69,
              "loc": Object {
                "end": Object {
                  "column": 32,
                  "line": 69,
                  "offset": 1732,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 69,
                  "offset": 1727,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check4: Store<string> = e",
          "descr": "Cannot assign \`e\` to \`check4\` because number [1] is incompatible with string [2] in type argument \`State\` [3].",
          "end": 35,
          "endline": 34,
          "line": 34,
          "loc": Object {
            "end": Object {
              "column": 35,
              "line": 34,
              "offset": 636,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 34,
              "offset": 635,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const f = createEffect<string, any, any>()",
              "descr": "[1]",
              "end": 33,
              "endline": 36,
              "line": 36,
              "loc": Object {
                "end": Object {
                  "column": 33,
                  "line": 36,
                  "offset": 671,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 28,
                  "line": 36,
                  "offset": 665,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check6: Event<number> = g",
              "descr": "[2]",
              "end": 30,
              "endline": 40,
              "line": 40,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 40,
                  "offset": 779,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 40,
                  "offset": 773,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Event<Payload> implements Unit<Payload> {",
              "descr": "[3]",
              "end": 34,
              "endline": 29,
              "line": 29,
              "loc": Object {
                "end": Object {
                  "column": 34,
                  "line": 29,
                  "offset": 486,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 29,
                  "offset": 479,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check6: Event<number> = g",
          "descr": "Cannot assign \`g\` to \`check6\` because string [1] is incompatible with number [2] in type argument \`Payload\` [3].",
          "end": 35,
          "endline": 40,
          "line": 40,
          "loc": Object {
            "end": Object {
              "column": 35,
              "line": 40,
              "offset": 784,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 40,
              "offset": 783,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check1: Event<string> = computed",
              "descr": "[1]",
              "end": 30,
              "endline": 52,
              "line": 52,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 52,
                  "offset": 1048,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 52,
                  "offset": 1042,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check2: Event<number> = computed",
              "descr": "[2]",
              "end": 30,
              "endline": 53,
              "line": 53,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 53,
                  "offset": 1091,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 53,
                  "offset": 1085,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Event<Payload> implements Unit<Payload> {",
              "descr": "[3]",
              "end": 34,
              "endline": 29,
              "line": 29,
              "loc": Object {
                "end": Object {
                  "column": 34,
                  "line": 29,
                  "offset": 486,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 29,
                  "offset": 479,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check2: Event<number> = computed",
          "descr": "Cannot assign \`computed\` to \`check2\` because string [1] is incompatible with number [2] in type argument \`Payload\` [3].",
          "end": 42,
          "endline": 53,
          "line": 53,
          "loc": Object {
            "end": Object {
              "column": 42,
              "line": 53,
              "offset": 1103,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 53,
              "offset": 1095,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const store2: Store<string> = createStore(0)",
              "descr": "[1]",
              "end": 47,
              "endline": 68,
              "line": 68,
              "loc": Object {
                "end": Object {
                  "column": 47,
                  "line": 68,
                  "offset": 1421,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 47,
                  "line": 68,
                  "offset": 1420,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 47,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const store2: Store<string> = createStore(0)",
              "descr": "[2]",
              "end": 30,
              "endline": 68,
              "line": 68,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 68,
                  "offset": 1404,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 68,
                  "offset": 1398,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Store<State> implements Unit<State> {",
              "descr": "[3]",
              "end": 32,
              "endline": 69,
              "line": 69,
              "loc": Object {
                "end": Object {
                  "column": 32,
                  "line": 69,
                  "offset": 1732,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 69,
                  "offset": 1727,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const store2: Store<string> = createStore(0)",
          "descr": "Cannot assign \`createStore(...)\` to \`store2\` because number [1] is incompatible with string [2] in type argument \`State\` [3].",
          "end": 47,
          "endline": 68,
          "line": 68,
          "loc": Object {
            "end": Object {
              "column": 47,
              "line": 68,
              "offset": 1421,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 47,
              "line": 68,
              "offset": 1420,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 47,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check1: Store<string> = computed",
              "descr": "[1]",
              "end": 30,
              "endline": 117,
              "line": 117,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 117,
                  "offset": 2752,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 117,
                  "offset": 2746,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const check2: Store<number> = computed",
              "descr": "[2]",
              "end": 30,
              "endline": 119,
              "line": 119,
              "loc": Object {
                "end": Object {
                  "column": 30,
                  "line": 119,
                  "offset": 2796,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 25,
                  "line": 119,
                  "offset": 2790,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 25,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export class Store<State> implements Unit<State> {",
              "descr": "[3]",
              "end": 32,
              "endline": 69,
              "line": 69,
              "loc": Object {
                "end": Object {
                  "column": 32,
                  "line": 69,
                  "offset": 1732,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 28,
                  "line": 69,
                  "offset": 1727,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 28,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const check2: Store<number> = computed",
          "descr": "Cannot assign \`computed\` to \`check2\` because string [1] is incompatible with number [2] in type argument \`State\` [3].",
          "end": 42,
          "endline": 119,
          "line": 119,
          "loc": Object {
            "end": Object {
              "column": 42,
              "line": 119,
              "offset": 2808,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 35,
              "line": 119,
              "offset": 2800,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 35,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const domain3 = createDomain(234)",
              "descr": "[1]",
              "end": 36,
              "endline": 186,
              "line": 186,
              "loc": Object {
                "end": Object {
                  "column": 36,
                  "line": 186,
                  "offset": 4573,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 34,
                  "line": 186,
                  "offset": 4570,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 34,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export function createDomain(domainName?: string): Domain",
              "descr": "[2]",
              "end": 56,
              "endline": 323,
              "line": 323,
              "loc": Object {
                "end": Object {
                  "column": 56,
                  "line": 323,
                  "offset": 7993,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 51,
                  "line": 323,
                  "offset": 7987,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 51,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const domain3 = createDomain(234)",
          "descr": "Cannot call \`createDomain\` with \`234\` bound to \`domainName\` because number [1] is incompatible with string [2].",
          "end": 36,
          "endline": 186,
          "line": 186,
          "loc": Object {
            "end": Object {
              "column": 36,
              "line": 186,
              "offset": 4573,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 34,
              "line": 186,
              "offset": 4570,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 34,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
    Object {
      "extra": Array [
        Object {
          "message": Array [
            Object {
              "context": null,
              "descr": "References:",
              "end": 0,
              "endline": 0,
              "line": 0,
              "path": "",
              "start": 1,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "    const domain4 = createDomain({foo: true})",
              "descr": "[1]",
              "end": 44,
              "endline": 187,
              "line": 187,
              "loc": Object {
                "end": Object {
                  "column": 44,
                  "line": 187,
                  "offset": 4619,
                },
                "source": "types.test.js",
                "start": Object {
                  "column": 34,
                  "line": 187,
                  "offset": 4608,
                },
                "type": "SourceFile",
              },
              "path": "types.test.js",
              "start": 34,
              "type": "Blame",
            },
          ],
        },
        Object {
          "message": Array [
            Object {
              "context": "declare export function createDomain(domainName?: string): Domain",
              "descr": "[2]",
              "end": 56,
              "endline": 323,
              "line": 323,
              "loc": Object {
                "end": Object {
                  "column": 56,
                  "line": 323,
                  "offset": 7993,
                },
                "source": "../../packages/effector/index.js.flow",
                "start": Object {
                  "column": 51,
                  "line": 323,
                  "offset": 7987,
                },
                "type": "SourceFile",
              },
              "path": "../../packages/effector/index.js.flow",
              "start": 51,
              "type": "Blame",
            },
          ],
        },
      ],
      "kind": "infer",
      "level": "error",
      "message": Array [
        Object {
          "context": "    const domain4 = createDomain({foo: true})",
          "descr": "Cannot call \`createDomain\` with object literal bound to \`domainName\` because object literal [1] is incompatible with string [2].",
          "end": 44,
          "endline": 187,
          "line": 187,
          "loc": Object {
            "end": Object {
              "column": 44,
              "line": 187,
              "offset": 4619,
            },
            "source": "types.test.js",
            "start": Object {
              "column": 34,
              "line": 187,
              "offset": 4608,
            },
            "type": "SourceFile",
          },
          "path": "types.test.js",
          "start": 34,
          "type": "Blame",
        },
      ],
      "suppressions": Array [],
    },
  ],
  "flowVersion": "0.97.0",
  "jsonVersion": "1",
  "passed": false,
}
`;

exports[`TypeScript: rejected 1`] = `
"Command failed: npx tsc -p src/types

src/types/types.test.ts(28,11): error TS2322: Type 'Event<number>' is not assignable to type 'Event<string>'.
  Type 'number' is not assignable to type 'string'.
src/types/types.test.ts(34,11): error TS2322: Type 'Store<number>' is not assignable to type 'Store<string>'.
  Type 'number' is not assignable to type 'string'.
src/types/types.test.ts(40,11): error TS2322: Type 'Event<string>' is not assignable to type 'Event<number>'.
  Type 'string' is not assignable to type 'number'.
src/types/types.test.ts(53,11): error TS2322: Type 'Event<string>' is not assignable to type 'Event<number>'.
src/types/types.test.ts(68,11): error TS2322: Type 'Store<number>' is not assignable to type 'Store<string>'.
src/types/types.test.ts(119,11): error TS2322: Type 'Store<string>' is not assignable to type 'Store<number>'.
  Type 'string' is not assignable to type 'number'.
src/types/types.test.ts(186,34): error TS2345: Argument of type '234' is not assignable to parameter of type 'string | undefined'.
src/types/types.test.ts(187,34): error TS2345: Argument of type '{ foo: boolean; }' is not assignable to parameter of type 'string'.
"
`;
